// @buape/carbon/dist/src/structures/GuildMember.js
import { Base } from "../abstracts/Base.js";
import { Role } from "./Role.js";
import { User } from "./User.js";
export class GuildMember extends Base {
    constructor(client, rawData, guild) {
        super(client);
        this.rawData = rawData;
        this.guild = guild;
        this.user = new User(client, rawData.user);
        this.setData(rawData);
    }
    rawData = null;
    setData(data) {
        if (!data)
            throw new Error("Cannot set data without having data... smh");
        this.rawData = data;
    }
    setField(key, value) {
        if (!this.rawData)
            throw new Error("Cannot set field without having data... smh");
        Reflect.set(this.rawData, key, value);
    }
    /**
     * The guild object of the member.
     */
    guild;
    /**
     * The user object of the member.
     */
    user;
    /**
     * The guild-specific nickname of the member.
     */
    get nickname() {
        if (!this.rawData)
            return undefined;
        return this.rawData.nick ?? null;
    }
    /**
     * The guild-specific avatar hash of the member.
     * You can use {@link GuildMember.avatarUrl} to get the URL of the avatar.
     */
    get avatar() {
        if (!this.rawData)
            return undefined;
        return this.rawData.avatar ?? null;
    }
    /**
     * Get the URL of the member's guild-specific avatar
     */
    get avatarUrl() {
        if (!this.rawData)
            return undefined;
        if (!this.user || !this.avatar)
            return null;
        return `https://cdn.discordapp.com/guilds/${this.guild.id}/users/${this.user.id}/${this.avatar}.png`;
    }
    /**
     * Is this member muted in Voice Channels?
     */
    get mute() {
        if (!this.rawData)
            return undefined;
        return this.rawData.mute;
    }
    /**
     * Is this member deafened in Voice Channels?
     */
    get deaf() {
        if (!this.rawData)
            return undefined;
        return this.rawData.deaf;
    }
    /**
     * The date since this member boosted the guild, if applicable.
     */
    get premiumSince() {
        if (!this.rawData)
            return undefined;
        return this.rawData.premium_since ?? null;
    }
    /**
     * The flags of the member.
     * @see https://discord.com/developers/docs/resources/guild#guild-member-object-guild-member-flags
     */
    get flags() {
        if (!this.rawData)
            return undefined;
        return this.rawData.flags;
    }
    /**
     * The roles of the member
     */
    get roles() {
        if (!this.rawData)
            return undefined;
        const roles = this.rawData.roles ?? [];
        return roles.map((role) => new Role(this.client, role));
    }
    /**
     * The joined date of the member
     */
    get joinedAt() {
        if (!this.rawData)
            return undefined;
        return this.rawData.joined_at;
    }
    /**
     * The date when the member's communication privileges (timeout) will be reinstated
     */
    get communicationDisabledUntil() {
        if (!this.rawData)
            return undefined;
        return this.rawData.communication_disabled_until ?? null;
    }
    /**
     * Is this member yet to pass the guild's Membership Screening requirements?
     */
    get pending() {
        if (!this.rawData)
            return undefined;
        return this.rawData.pending ?? false;
    }
    /**
     * Set the nickname of the member
     */
    async setNickname(nickname) {
        await this.client.rest.patch(`/guilds/${this.guild?.id}/members/${this.user?.id}`, {
            body: {
                nick: nickname
            }
        });
        this.setField("nick", nickname);
    }
    /**
     * Add a role to the member
     */
    async addRole(roleId) {
        await this.client.rest.put(`/guilds/${this.guild?.id}/members/${this.user?.id}/roles/${roleId}`, {});
        this.roles?.push(new Role(this.client, roleId));
    }
    /**
     * Remove a role from the member
     */
    async removeRole(roleId) {
        await this.client.rest.delete(`/guilds/${this.guild?.id}/members/${this.user?.id}/roles/${roleId}`);
        const roles = this.roles?.filter((role) => role.id !== roleId);
        if (roles)
            this.setField("roles", roles);
    }
    /**
     * Kick the member from the guild
     */
    async kick() {
        await this.client.rest.delete(`/guilds/${this.guild?.id}/members/${this.user?.id}`);
    }
    /**
     * Ban the member from the guild
     */
    async ban(options = {}) {
        await this.client.rest.put(`/guilds/${this.guild?.id}/bans/${this.user?.id}`, {
            body: {
                reason: options.reason,
                delete_message_days: options.deleteMessageDays
            }
        });
    }
    /**
     * Mute a member in voice channels
     */
    async muteMember() {
        await this.client.rest.patch(`/guilds/${this.guild?.id}/members/${this.user?.id}`, {
            body: {
                mute: true
            }
        });
        this.setField("mute", true);
    }
    /**
     * Unmute a member in voice channels
     */
    async unmuteMember() {
        await this.client.rest.patch(`/guilds/${this.guild?.id}/members/${this.user?.id}`, {
            body: {
                mute: false
            }
        });
        this.setField("mute", false);
    }
    /**
     * Deafen a member in voice channels
     */
    async deafenMember() {
        await this.client.rest.patch(`/guilds/${this.guild?.id}/members/${this.user?.id}`, {
            body: {
                deaf: true
            }
        });
        this.setField("deaf", true);
    }
    /**
     * Undeafen a member in voice channels
     */
    async undeafenMember() {
        await this.client.rest.patch(`/guilds/${this.guild?.id}/members/${this.user?.id}`, {
            body: {
                deaf: false
            }
        });
        this.setField("deaf", false);
    }
    /**
     * Set or remove a timeout for a member in the guild
     */
    async timeoutMember(communicationDisabledUntil) {
        await this.client.rest.patch(`/guilds/${this.guild?.id}/members/${this.user?.id}`, {
            body: {
                communication_disabled_until: communicationDisabledUntil
            }
        });
        this.setField("communication_disabled_until", communicationDisabledUntil);
    }
}
