// @buape/carbon/dist/src/structures/Guild.js
import { Routes } from "discord-api-types/v10";
import { Base } from "../abstracts/Base.js";
import { channelFactory } from "../factories/channelFactory.js";
import { GuildMember } from "./GuildMember.js";
import { Role } from "./Role.js";
export class Guild extends Base {
    constructor(client, rawDataOrId) {
        super(client);
        if (typeof rawDataOrId === "string") {
            this.id = rawDataOrId;
        }
        else {
            this.rawData = rawDataOrId;
            this.id = rawDataOrId.id;
            this.setData(rawDataOrId);
        }
    }
    rawData = null;
    setData(data) {
        if (!data)
            throw new Error("Cannot set data without having data... smh");
        this.rawData = data;
    }
    // private setField(key: keyof APIGuild, value: unknown) {
    // 	if (!this.rawData)
    // 		throw new Error("Cannot set field without having data... smh")
    // 	Reflect.set(this.rawData, key, value)
    // }
    /**
     * The ID of the guild
     */
    id;
    /**
     * Whether the guild is a partial guild (meaning it does not have all the data).
     * If this is true, you should use {@link Guild.fetch} to get the full data of the guild.
     */
    get partial() {
        return (this.rawData === null);
    }
    /**
     * The name of the guild.
     */
    get name() {
        if (!this.rawData)
            return undefined;
        return this.rawData.name;
    }
    /**
     * The description of the guild.
     */
    get description() {
        if (!this.rawData)
            return undefined;
        return this.rawData.description;
    }
    /**
     * The icon hash of the guild.
     * You can use {@link Guild.iconUrl} to get the URL of the icon.
     */
    get icon() {
        if (!this.rawData)
            return undefined;
        return this.rawData.icon;
    }
    /**
     * Get the URL of the guild's icon
     */
    get iconUrl() {
        if (!this.rawData)
            return undefined;
        if (!this.icon)
            return null;
        return `https://cdn.discordapp.com/icons/${this.id}/${this.icon}.png`;
    }
    /**
     * The splash hash of the guild.
     * You can use {@link Guild.splashUrl} to get the URL of the splash.
     */
    get splash() {
        if (!this.rawData)
            return undefined;
        return this.rawData.splash;
    }
    /**
     * Get the URL of the guild's splash
     */
    get splashUrl() {
        if (!this.rawData)
            return undefined;
        if (!this.splash)
            return null;
        return `https://cdn.discordapp.com/splashes/${this.id}/${this.splash}.png`;
    }
    /**
     * The ID of the owner of the guild.
     */
    get ownerId() {
        if (!this.rawData)
            return undefined;
        return this.rawData.owner_id;
    }
    /**
     * Get all roles in the guild
     */
    get roles() {
        if (!this.rawData)
            return undefined;
        const roles = this.rawData?.roles;
        if (!roles)
            throw new Error("Cannot get roles without having data... smh");
        return roles.map((role) => new Role(this.client, role));
    }
    /**
     * Fetch updated data for this guild.
     * If the guild is partial, this will fetch all the data for the guild and populate the fields.
     * If the guild is not partial, all fields will be updated with new values from Discord.
     */
    async fetch() {
        const newData = (await this.client.rest.get(Routes.guild(this.id)));
        if (!newData)
            throw new Error(`Guild ${this.id}not found`);
        this.setData(newData);
    }
    /**
     * Leave the guild
     */
    async leave() {
        return await this.client.rest.delete(Routes.guild(this.id));
    }
    /**
     * Create a role in the guild
     */
    async createRole(data) {
        const role = (await this.client.rest.post(Routes.guildRoles(this.id), {
            body: {
                ...data
            }
        }));
        return new Role(this.client, role);
    }
    /**
     * Get a member in the guild by ID
     */
    async fetchMember(memberId) {
        const member = (await this.client.rest.get(Routes.guildMember(this.id, memberId)));
        return new GuildMember(this.client, member, this);
    }
    /**
     * Fetch a channel from the guild by ID
     */
    async fetchChannel(channelId) {
        const channel = (await this.client.rest.get(Routes.channel(channelId)));
        return channelFactory(this.client, channel);
    }
}
