// @buape/carbon/dist/src/structures/Message.js
import { Routes } from "discord-api-types/v10";
import { Base } from "../abstracts/Base.js";
import { Embed } from "../classes/Embed.js";
import { channelFactory } from "../factories/channelFactory.js";
import { GuildThreadChannel } from "./GuildThreadChannel.js";
import { Role } from "./Role.js";
import { User } from "./User.js";
export class Message extends Base {
    constructor(client, rawDataOrIds) {
        super(client);
        if (Object.keys(rawDataOrIds).length === 2) {
            this.id = rawDataOrIds.id;
            this.channelId = rawDataOrIds["channelId"];
        }
        else {
            const data = rawDataOrIds;
            this.id = data.id;
            this.channelId = data.channel_id;
            this.setData(data);
        }
    }
    rawData = null;
    setData(data) {
        this.rawData = data;
        if (!data)
            throw new Error("Cannot set data without having data... smh");
    }
    // private setField(key: keyof APIMessage, value: unknown) {
    // 	if (!this.rawData)
    // 		throw new Error("Cannot set field without having data... smh")
    // 	Reflect.set(this.rawData, key, value)
    // }
    /**
     * The ID of the message
     */
    id;
    /**
     * The ID of the channel the message is in
     */
    channelId;
    /**
     * Whether the message is a partial message (meaning it does not have all the data).
     * If this is true, you should use {@link Message.fetch} to get the full data of the message.
     */
    get partial() {
        return (this.rawData === null);
    }
    /**
     * If this message is a response to an interaction, this is the ID of the interaction's application
     */
    get applicationId() {
        if (!this.rawData)
            return undefined;
        return this.rawData.application_id;
    }
    /**
     * The attachments of the message
     */
    get attachments() {
        if (!this.rawData)
            return undefined;
        return this.rawData.attachments ?? [];
    }
    /**
     * The components of the message
     */
    get components() {
        if (!this.rawData)
            return undefined;
        return this.rawData.components ?? [];
    }
    /**
     * The content of the message
     */
    get content() {
        if (!this.rawData)
            return undefined;
        return this.rawData.content ?? "";
    }
    get embeds() {
        if (!this.rawData)
            return undefined;
        if (!this.rawData?.embeds)
            return [];
        return this.rawData.embeds.map((embed) => new Embed(embed));
    }
    /**
     * If this message was edited, this is the timestamp of the edit
     */
    get editedTimestamp() {
        if (!this.rawData)
            return undefined;
        return this.rawData.edited_timestamp;
    }
    /**
     * The flags of the message
     */
    get flags() {
        if (!this.rawData)
            return undefined;
        return this.rawData.flags;
    }
    /**
     * The interaction metadata of the message
     */
    get interactionMetadata() {
        if (!this.rawData)
            return undefined;
        return this.rawData.interaction_metadata;
    }
    /**
     * Whether the message mentions everyone
     */
    get mentionedEveryone() {
        if (!this.rawData)
            return undefined;
        return this.rawData.mention_everyone;
    }
    /**
     * The users mentioned in the message
     */
    get mentionedUsers() {
        if (!this.rawData)
            return undefined;
        if (!this.rawData?.mentions)
            return [];
        return this.rawData.mentions.map((mention) => new User(this.client, mention));
    }
    /**
     * The roles mentioned in the message
     */
    get mentionedRoles() {
        if (!this.rawData)
            return undefined;
        if (!this.rawData?.mention_roles)
            return [];
        return this.rawData.mention_roles.map((mention) => new Role(this.client, mention));
    }
    /**
     * The data about the referenced message. You can use {@link Message.referencedMessage} to get the referenced message itself.
     */
    get messageReference() {
        if (!this.rawData)
            return undefined;
        return this.rawData.message_reference;
    }
    /**
     * The referenced message itself
     */
    get referencedMessage() {
        if (!this.rawData?.referenced_message)
            return null;
        return new Message(this.client, this.rawData?.referenced_message);
    }
    /**
     * Whether the message is pinned
     */
    get pinned() {
        if (!this.rawData)
            return undefined;
        return this.rawData.pinned;
    }
    /**
     * The poll contained in the message
     */
    get poll() {
        if (!this.rawData)
            return undefined;
        return this.rawData.poll;
    }
    /**
     * The approximate position of the message in the channel
     */
    get position() {
        if (!this.rawData)
            return undefined;
        return this.rawData.position;
    }
    /**
     * The reactions on the message
     */
    get reactions() {
        if (!this.rawData)
            return undefined;
        return this.rawData.reactions ?? [];
    }
    /**
     * The stickers in the message
     */
    get stickers() {
        if (!this.rawData)
            return undefined;
        return this.rawData.sticker_items ?? [];
    }
    /**
     * The timestamp of the original message
     */
    get timestamp() {
        if (!this.rawData)
            return undefined;
        return this.rawData.timestamp;
    }
    /**
     * Whether the message is a TTS message
     */
    get tts() {
        if (!this.rawData)
            return undefined;
        return this.rawData.tts;
    }
    /**
     * The type of the message
     */
    get type() {
        if (!this.rawData)
            return undefined;
        return this.rawData.type;
    }
    /**
     * Get the author of the message
     */
    get author() {
        if (!this.rawData)
            return null;
        if (this.rawData?.webhook_id)
            return null; // TODO: Add webhook user
        return new User(this.client, this.rawData.author);
    }
    /**
     * Get the thread associated with this message, if there is one
     */
    get thread() {
        if (!this.rawData)
            return null;
        if (!this.rawData?.thread)
            return null;
        return channelFactory(this.client, this.rawData?.thread);
    }
    /**
     * Fetch updated data for this message.
     * If the message is partial, this will fetch all the data for the message and populate the fields.
     * If the message is not partial, all fields will be updated with new values from Discord.
     */
    async fetch() {
        const newData = (await this.client.rest.get(Routes.channelMessage(this.channelId, this.id)));
        if (!newData)
            throw new Error(`Message ${this.id} not found`);
        this.setData(newData);
    }
    /**
     * Delete this message from Discord
     */
    async delete() {
        return await this.client.rest.delete(Routes.channelMessage(this.channelId, this.id));
    }
    /**
     * Get the channel the message was sent in
     */
    async fetchChannel() {
        const data = (await this.client.rest.get(Routes.channel(this.channelId)));
        return channelFactory(this.client, data);
    }
    /**
     * Pin this message
     */
    async pin() {
        await this.client.rest.put(Routes.channelPin(this.channelId, this.id));
    }
    /**
     * Unpin this message
     */
    async unpin() {
        await this.client.rest.delete(Routes.channelPin(this.channelId, this.id));
    }
    /**
     * Start a thread with this message as the associated start message.
     * If you want to start a thread without a start message, use {@link BaseGuildTextChannel.startThread}
     */
    async startThread(data) {
        const thread = (await this.client.rest.post(Routes.threads(this.channelId, this.id), {
            body: { ...data }
        }));
        return new GuildThreadChannel(this.client, thread);
    }
    async edit(data) {
        await this.client.rest.patch(Routes.channelMessage(this.channelId, this.id), {
            body: {
                ...data,
                embeds: data.embeds?.map((embed) => embed.serialize()),
                components: data.components?.map((row) => row.serialize()),
                allowed_mentions: data.allowedMentions
            }
        });
    }
}
