// @buape/carbon/dist/src/structures/Role.js
import { Routes } from "discord-api-types/v10";
import { Base } from "../abstracts/Base.js";
export class Role extends Base {
    constructor(client, rawDataOrId) {
        super(client);
        if (typeof rawDataOrId === "string") {
            this.id = rawDataOrId;
        }
        else {
            this.rawData = rawDataOrId;
            this.id = rawDataOrId.id;
            this.setData(rawDataOrId);
        }
    }
    rawData = null;
    setData(data) {
        if (!data)
            throw new Error("Cannot set data without having data... smh");
        this.rawData = data;
    }
    setField(key, value) {
        if (!this.rawData)
            throw new Error("Cannot set field without having data... smh");
        Reflect.set(this.rawData, key, value);
    }
    /**
     * The ID of the role.
     */
    id;
    /**
     * Whether the role is a partial role (meaning it does not have all the data).
     * If this is true, you should use {@link Role.fetch} to get the full data of the role.
     */
    get partial() {
        return (this.rawData === null);
    }
    /**
     * The name of the role.
     */
    get name() {
        if (!this.rawData)
            return undefined;
        return this.rawData.name;
    }
    /**
     * The color of the role.
     */
    get color() {
        if (!this.rawData)
            return undefined;
        return this.rawData.color;
    }
    /**
     * The icon hash of the role.
     * You can use {@link Role.iconUrl} to get the URL of the icon.
     */
    get icon() {
        if (!this.rawData)
            return undefined;
        return this.rawData.icon ?? null;
    }
    /**
     * Get the URL of the role's icon
     */
    get iconUrl() {
        if (!this.rawData)
            return undefined;
        if (!this.icon)
            return null;
        return `https://cdn.discordapp.com/role-icons/${this.id}/${this.icon}.png`;
    }
    /**
     * If this role is mentionable.
     */
    get mentionable() {
        if (!this.rawData)
            return undefined;
        return this.rawData.mentionable;
    }
    /**
     * If this role is hoisted.
     */
    get hoisted() {
        if (!this.rawData)
            return undefined;
        return this.rawData.hoist;
    }
    /**
     * The position of the role.
     */
    get position() {
        if (!this.rawData)
            return undefined;
        return this.rawData.position;
    }
    /**
     * The permissions of the role.
     */
    get permissions() {
        if (!this.rawData)
            return undefined;
        return this.rawData.permissions;
    }
    /**
     * If this role is managed by an integration.
     */
    get managed() {
        if (!this.rawData)
            return undefined;
        return this.rawData.managed;
    }
    /**
     * The unicode emoji for the role.
     */
    get unicodeEmoji() {
        if (!this.rawData)
            return undefined;
        return this.rawData.unicode_emoji ?? null;
    }
    /**
     * The flags of this role.
     * @see https://discord.com/developers/docs/topics/permissions#role-object-role-flags
     */
    get flags() {
        if (!this.rawData)
            return undefined;
        return this.rawData.flags;
    }
    /**
     * The tags of this role.
     * @see https://discord.com/developers/docs/topics/permissions#role-object-role-tags-structure
     */
    get tags() {
        if (!this.rawData)
            return undefined;
        return this.rawData.tags;
    }
    /**
     * Fetch updated data for this role.
     * If the role is partial, this will fetch all the data for the role and populate the fields.
     * If the role is not partial, all fields will be updated with new values from Discord.
     */
    async fetch(guildId) {
        const newData = (await this.client.rest.get(Routes.guildRole(guildId, this.id)));
        if (!newData)
            throw new Error(`Role ${this.id} not found`);
        this.setData(newData);
    }
    /**
     * Set the name of the role
     */
    async setName(guildId, name) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: {
                name
            }
        });
        this.setField("name", name);
    }
    /**
     * Set the color of the role
     */
    async setColor(guildId, color) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: { color }
        });
        this.setField("color", color);
    }
    /**
     * Set the icon of the role
     * @param icon The unicode emoji or icon URL to set
     */
    async setIcon(guildId, icon) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: { icon }
        });
        this.setField("icon", icon);
    }
    /**
     * Set the mentionable status of the role
     */
    async setMentionable(guildId, mentionable) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: { mentionable }
        });
        this.setField("mentionable", mentionable);
    }
    /**
     * Set the hoisted status of the role
     */
    async setHoisted(guildId, hoisted) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: { hoist: hoisted }
        });
        this.setField("hoist", hoisted);
    }
    /**
     * Set the position of the role
     */
    async setPosition(guildId, position) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: { position }
        });
        this.setField("position", position);
    }
    /**
     * Set the permissions of the role
     * @param permissions The permissions to set as a BitField string, until a better permission structure is implemented
     */
    async setPermissions(guildId, permissions) {
        await this.client.rest.patch(Routes.guildRole(guildId, this.id), {
            body: { permissions }
        });
        this.setField("permissions", permissions);
    }
    async delete(guildId) {
        await this.client.rest.delete(Routes.guildRole(guildId, this.id));
    }
}
