// @buape/carbon/dist/src/classes/Client.js
import { RequestClient } from "@buape/carbon-request";
import { InteractionResponseType, InteractionType, Routes } from "discord-api-types/v10";
import { AutoRouter, StatusError, json } from "itty-router";
import { channelFactory } from "../factories/channelFactory.js";
import { CommandHandler } from "../internals/CommandHandler.js";
import { ComponentHandler } from "../internals/ComponentHandler.js";
import { ModalHandler } from "../internals/ModalHandler.js";
import { Guild } from "../structures/Guild.js";
import { GuildMember } from "../structures/GuildMember.js";
import { Role } from "../structures/Role.js";
import { User } from "../structures/User.js";
import { concatUint8Arrays, subtleCrypto, valueToUint8Array } from "../utils.js";
/**
 * The mode that the client is running in.
 * Different platforms have different requirements for how processes are handled.
 */
export var ClientMode;
(function (ClientMode) {
    ClientMode["NodeJS"] = "node";
    ClientMode["CloudflareWorkers"] = "cloudflare";
    ClientMode["Bun"] = "bun";
    ClientMode["Vercel"] = "vercel";
    ClientMode["Web"] = "web";
})(ClientMode || (ClientMode = {}));
/**
 * The main client used to interact with Discord
 */
export class Client {
    /**
     * The options used to initialize the client
     */
    options;
    /**
     * The commands that the client has registered
     */
    commands;
    /**
     * The router used to handle requests
     */
    router;
    /**
     * The rest client used to interact with the Discord API
     */
    rest;
    /**
     * The handler for the component interactions sent from Discord
     * @internal
     */
    componentHandler;
    /**
     * The handler for the modal interactions sent from Discord
     * @internal
     */
    commandHandler;
    /**
     * The handler for the modal interactions sent from Discord
     * @internal
     */
    modalHandler;
    /**
     * Creates a new client
     * @param options The options used to initialize the client
     * @param commands The commands that the client has registered
     */
    constructor(options, commands) {
        if (!options.clientId)
            throw new Error("Missing client ID");
        if (!options.publicKey)
            throw new Error("Missing public key");
        if (!options.token)
            throw new Error("Missing token");
        this.options = options;
        this.commands = commands;
        this.commandHandler = new CommandHandler(this);
        this.componentHandler = new ComponentHandler(this);
        this.modalHandler = new ModalHandler(this);
        const routerData = this.options.mode === ClientMode.Bun && this.options.port
            ? { port: this.options.port }
            : {};
        // biome-ignore lint/suspicious/noExplicitAny: <explanation>
        this.router = AutoRouter(routerData);
        this.rest = new RequestClient(options.token, options.requestOptions);
        if (this.options.autoRegister) {
            for (const command of commands) {
                for (const component of command.components)
                    this.componentHandler.registerComponent(new component());
                for (const modal of command.modals)
                    this.modalHandler.registerModal(new modal());
            }
        }
        if (this.options.autoDeploy)
            this.deployCommands();
        this.setupRoutes();
    }
    /**
     * Deploy the commands registered to Discord.
     * This is automatically called when running in NodeJS mode.
     */
    async deployCommands() {
        try {
            const commands = this.commands
                .filter((x) => x.name !== "*")
                .map((command) => {
                return command.serialize();
            });
            await this.rest.put(Routes.applicationCommands(this.options.clientId), {
                body: commands
            });
            console.log(`Deployed ${commands.length} commands to Discord`);
        }
        catch (err) {
            console.error("Failed to deploy commands");
            console.error(err);
        }
    }
    /**
     * Setup the routes for the client
     */
    setupRoutes() {
        this.router.get("/", () => {
            if (this.options.redirectUrl)
                return Response.redirect(this.options.redirectUrl, 302);
            throw new StatusError(404);
        });
        this.router.post(this.options.interactionRoute || "/interaction", async (req, ctx) => {
            return await this.handle(req, ctx);
        });
    }
    /**
     * If you want use a custom handler for HTTP requests instead of Carbon's router, you can use this method.
     * @param req The request to handle
     * @param ctx Cloudflare Workers only. The execution context of the request, provided in the fetch handler from CF.
     * @returns A response to send back to the client.
     */
    async handle(req, ctx) {
        const isValid = await this.validateInteraction(req);
        if (!isValid) {
            return new Response("Invalid request signature", { status: 401 });
        }
        const rawInteraction = (await req.json());
        if (rawInteraction.type === InteractionType.Ping) {
            return json({
                type: InteractionResponseType.Pong
            });
        }
        if (rawInteraction.type === InteractionType.ApplicationCommand) {
            if (ctx?.waitUntil) {
                ctx.waitUntil((async () => {
                    await this.commandHandler.handleCommandInteraction(rawInteraction);
                })());
            }
            else {
                await this.commandHandler.handleCommandInteraction(rawInteraction);
            }
        }
        if (rawInteraction.type === InteractionType.ApplicationCommandAutocomplete) {
            if (ctx?.waitUntil) {
                ctx.waitUntil((async () => {
                    await this.commandHandler.handleAutocompleteInteraction(rawInteraction);
                })());
            }
            else {
                await this.commandHandler.handleAutocompleteInteraction(rawInteraction);
            }
        }
        if (rawInteraction.type === InteractionType.MessageComponent) {
            if (ctx?.waitUntil) {
                ctx.waitUntil((async () => {
                    await this.componentHandler.handleInteraction(rawInteraction);
                })());
            }
            else {
                await this.componentHandler.handleInteraction(rawInteraction);
            }
        }
        if (rawInteraction.type === InteractionType.ModalSubmit) {
            if (ctx?.waitUntil) {
                ctx.waitUntil((async () => {
                    await this.modalHandler.handleInteraction(rawInteraction);
                })());
            }
            else {
                await this.modalHandler.handleInteraction(rawInteraction);
            }
        }
        return new Response(null, { status: 202 });
    }
    /**
     * Validate the interaction request
     * @param req The request to validate
     */
    async validateInteraction(req) {
        const body = await req.clone().text();
        const signature = req.headers.get("X-Signature-Ed25519");
        const timestamp = req.headers.get("X-Signature-Timestamp");
        if (!timestamp || !signature || req.method !== "POST" || !body) {
            throw new StatusError(401);
        }
        try {
            const timestampData = valueToUint8Array(timestamp);
            const bodyData = valueToUint8Array(body);
            const message = concatUint8Arrays(timestampData, bodyData);
            const isValid = await subtleCrypto.verify({
                name: "ed25519"
            }, await subtleCrypto.importKey("raw", valueToUint8Array(this.options.publicKey, "hex"), {
                name: "ed25519",
                namedCurve: "ed25519"
            }, false, ["verify"]), valueToUint8Array(signature, "hex"), message);
            return isValid;
        }
        catch (_) {
            return false;
        }
    }
    // ======================== Begin Fetchers ================================================
    /**
     * Fetch a user from the Discord API
     * @param id The ID of the user to fetch
     * @returns The user data
     */
    async fetchUser(id) {
        const user = (await this.rest.get(Routes.user(id)));
        return new User(this, user);
    }
    /**
     * Fetch a guild from the Discord API
     * @param id The ID of the guild to fetch
     * @returns The guild data
     */
    async fetchGuild(id) {
        const guild = (await this.rest.get(Routes.guild(id)));
        return new Guild(this, guild);
    }
    /**
     * Fetch a channel from the Discord API
     * @param id The ID of the channel to fetch
     * @returns The channel data
     */
    async fetchChannel(id) {
        const channel = (await this.rest.get(Routes.channel(id)));
        return channelFactory(this, channel);
    }
    /**
     * Fetch a role from the Discord API
     * @param id The ID of the role to fetch
     * @returns The role data
     */
    async fetchRole(guildId, id) {
        const role = (await this.rest.get(Routes.guildRole(guildId, id)));
        return new Role(this, role);
    }
    /**
     * Fetch a member from the Discord API
     * @param guildId The ID of the guild the member is in
     * @param id The ID of the member to fetch
     * @returns The member data
     */
    async fetchMember(guildId, id) {
        const member = (await this.rest.get(Routes.guildMember(guildId, id)));
        return new GuildMember(this, member, new Guild(this, guildId));
    }
}
