// @buape/carbon/dist/src/classes/CommandWithSubcommandGroups.js
import { ApplicationCommandOptionType } from "discord-api-types/v10";
import { CommandWithSubcommands } from "./CommandWithSubcommands.js";
/**
 * Represents a subcommand group command that the user creates.
 * You make this instead of a {@link Command} class when you want to have subcommand groups in your options.
 */
export class CommandWithSubcommandGroups extends CommandWithSubcommands {
    /**
     * The subcommands that the user can use
     */
    subcommands = [];
    /**
     * @internal
     */
    serializeOptions() {
        const subcommands = this.subcommands.map((subcommand) => ({
            name: subcommand.name,
            description: subcommand.description,
            type: ApplicationCommandOptionType.Subcommand,
            options: subcommand.serializeOptions()
        }));
        const subcommandGroups = this.subcommandGroups.map((subcommandGroup) => ({
            name: subcommandGroup.name,
            description: subcommandGroup.description,
            type: ApplicationCommandOptionType.SubcommandGroup,
            options: subcommandGroup.serializeOptions()
        }));
        return [...subcommands, ...subcommandGroups];
    }
}
