// @buape/carbon/dist/src/classes/TextInput.js
import { ComponentType, TextInputStyle } from "discord-api-types/v10";
import { BaseComponent } from "../abstracts/BaseComponent.js";
export class TextInput extends BaseComponent {
    type = ComponentType.TextInput;
    /**
     * The style of the text input
     * @default TextInputStyle.Short
     */
    style = TextInputStyle.Short;
    /**
     * The minimum length of the text input
     */
    minLength;
    /**
     * The maximum length of the text input
     */
    maxLength;
    /**
     * Whether the text input is required
     */
    required;
    /**
     * The value of the text input
     */
    value;
    /**
     * The placeholder of the text input
     */
    placeholder;
    serialize = () => {
        return {
            type: ComponentType.TextInput,
            custom_id: this.customId,
            style: this.style,
            label: this.label,
            min_length: this.minLength,
            max_length: this.maxLength,
            required: this.required,
            value: this.value,
            placeholder: this.placeholder
        };
    };
}
