// @buape/carbon/dist/src/abstracts/BaseComponentInteraction.js
import { InteractionResponseType, Routes } from "discord-api-types/v10";
import { serializePayload, splitCustomId } from "../utils.js";
import { BaseInteraction } from "./BaseInteraction.js";
export class BaseComponentInteraction extends BaseInteraction {
    customId;
    componentType;
    constructor(client, data, defaults) {
        super(client, data, defaults);
        if (!data.data)
            throw new Error("Invalid interaction data was used to create this class");
        this.customId = splitCustomId(data.data.custom_id)[0];
        this.componentType = data.data.component_type;
    }
    /**
     * Acknowledge the interaction, the user does not see a loading state.
     * This can only be used for component interactions
     */
    async acknowledge() {
        await this.client.rest.post(Routes.interactionCallback(this.rawData.id, this.rawData.token), {
            body: {
                type: InteractionResponseType.DeferredMessageUpdate
            }
        });
        this._deferred = true;
    }
    /**
     * Update the original message of the component
     */
    async update(data, options = {}) {
        const serialized = serializePayload(data);
        await this.client.rest.post(Routes.interactionCallback(this.rawData.id, this.rawData.token), {
            body: {
                type: InteractionResponseType.UpdateMessage,
                data: {
                    ...serialized
                }
            },
            files: options.files
        });
    }
}
