// @buape/carbon/dist/src/abstracts/BaseChannel.js
import { Routes } from "discord-api-types/v10";
import { Base } from "./Base.js";
export class BaseChannel extends Base {
    constructor(client, rawDataOrId) {
        super(client);
        if (typeof rawDataOrId === "string") {
            this.id = rawDataOrId;
        }
        else {
            this.rawData = rawDataOrId;
            this.id = rawDataOrId.id;
            this.setData(rawDataOrId);
        }
    }
    /**
     * The raw data of the channel.
     */
    rawData = null;
    setData(data) {
        if (!data)
            throw new Error("Cannot set data without having data... smh");
        this.rawData = data;
    }
    setField(field, value) {
        if (!this.rawData)
            throw new Error("Cannot set field without having data... smh");
        this.rawData[field] = value;
    }
    /**
     * The id of the channel.
     */
    id;
    /**
     * Whether the channel is a partial channel (meaning it does not have all the data).
     * If this is true, you should use {@link BaseChannel.fetch} to get the full data of the channel.
     */
    get partial() {
        return (this.rawData === null);
    }
    /**
     * The type of the channel.
     */
    get type() {
        if (!this.rawData)
            return undefined;
        return this.rawData.type;
    }
    /**
     * The flags of the channel in a bitfield.
     * @see https://discord.com/developers/docs/resources/channel#channel-object-channel-flags
     */
    get flags() {
        if (!this.rawData)
            return undefined;
        return this.rawData.flags;
    }
    /**
     * Fetches the channel from the API.
     * @returns The channel data.
     */
    async fetch() {
        const newData = (await this.client.rest.get(Routes.channel(this.id)));
        if (!newData)
            throw new Error(`Channel ${this.id} not found`);
        this.setData(newData);
    }
}
