// @buape/carbon/dist/src/abstracts/BaseInteraction.js
import { InteractionResponseType, Routes } from "discord-api-types/v10";
import { Base, Embed, Guild, Message, User, channelFactory } from "../index.js";
import { GuildMember } from "../structures/GuildMember.js";
import { serializePayload } from "../utils.js";
/**
 * This is the base type interaction, all interaction types extend from this
 */
export class BaseInteraction extends Base {
    /**
     * The type of interaction
     */
    type;
    /**
     * The raw data of the interaction
     */
    rawData;
    /**
     * The user who sent the interaction
     */
    userId;
    /**
     * Whether the interaction is deferred already
     * @internal
     */
    _deferred = false;
    defaultEphemeral = false;
    constructor(client, data, defaults) {
        super(client);
        this.rawData = data;
        this.type = data.type;
        this.userId =
            this.rawData.user?.id || this.rawData.member?.user.id || undefined;
        if (defaults.ephemeral)
            this.defaultEphemeral = defaults.ephemeral;
    }
    get embeds() {
        if (!this.rawData.message)
            return null;
        return this.rawData.message.embeds.map((embed) => new Embed(embed));
    }
    get message() {
        if (!this.rawData.message)
            return null;
        return new Message(this.client, this.rawData.message);
    }
    get guild() {
        if (!this.rawData.guild_id)
            return null;
        return new Guild(this.client, this.rawData.guild_id);
    }
    get user() {
        if (this.rawData.user)
            return new User(this.client, this.rawData.user);
        if (this.rawData.member)
            return new User(this.client, this.rawData.member.user);
        return null;
    }
    get channel() {
        if (!this.rawData.channel)
            return null;
        return channelFactory(this.client, this.rawData.channel);
    }
    get member() {
        if (!this.rawData.member)
            return null;
        if (!this.guild)
            return null;
        return new GuildMember(this.client, this.rawData.member, this.guild);
    }
    /**
     * Reply to an interaction.
     * If the interaction is deferred, this will edit the original response.
     * @param data The response data
     */
    async reply(data, options = {}) {
        const serialized = serializePayload(data);
        if (this._deferred) {
            await this.client.rest.patch(Routes.webhookMessage(this.client.options.clientId, this.rawData.token, "@original"), {
                body: {
                    ...serialized
                },
                files: options.files
            });
        }
        else {
            await this.client.rest.post(Routes.interactionCallback(this.rawData.id, this.rawData.token), {
                body: {
                    type: InteractionResponseType.ChannelMessageWithSource,
                    data: {
                        ...serializePayload(data),
                        flags: options.ephemeral
                            ? 64 | ("flags" in serialized ? (serialized.flags ?? 0) : 0)
                            : undefined
                    }
                },
                files: options.files
            });
        }
    }
    /**
     * Defer the interaction response. This is used automatically by commands that are set to defer.
     * If the interaction is already deferred, this will do nothing.
     * @internal
     */
    async defer() {
        if (this._deferred)
            return;
        this._deferred = true;
        await this.client.rest.post(Routes.interactionCallback(this.rawData.id, this.rawData.token), {
            body: {
                type: InteractionResponseType.DeferredChannelMessageWithSource,
                flags: this.defaultEphemeral ? 64 : undefined
            }
        });
    }
    /**
     * Show a modal to the user
     * This can only be used if the interaction is not deferred
     */
    async showModal(modal) {
        if (this._deferred)
            throw new Error("You cannot defer an interaction that shows a modal");
        await this.client.rest.post(Routes.interactionCallback(this.rawData.id, this.rawData.token), {
            body: {
                type: InteractionResponseType.Modal,
                data: modal.serialize()
            }
        });
    }
    /**
     * Send a followup message to the interaction
     */
    async followUp(reply, options = {}) {
        const serialized = serializePayload(reply);
        await this.client.rest.post(Routes.webhook(this.client.options.clientId, this.rawData.token), {
            body: {
                ...serialized,
                flags: options.ephemeral
                    ? 64 | ("flags" in serialized ? (serialized.flags ?? 0) : 0)
                    : undefined
            },
            files: options.files
        });
    }
}
