// @buape/carbon/dist/src/abstracts/BaseCommand.js
import { ApplicationCommandType } from "discord-api-types/v10";
import { ApplicationIntegrationType, InteractionContextType } from "../index.js";
/**
 * Represents the base data of a command that the user creates
 */
export class BaseCommand {
    /**
     * Whether the command response should be automatically deferred
     */
    defer = false;
    /**
     * Whether the command response should be ephemeral
     */
    ephemeral = false;
    /**
     * The places this command can be used in
     * @beta API types are not finalized
     */
    integrationTypes = [
        ApplicationIntegrationType.GuildInstall,
        ApplicationIntegrationType.UserInstall
    ];
    /**
     * The contexts this command can be used in
     * @beta API types are not finalized
     */
    contexts = [
        InteractionContextType.Guild,
        InteractionContextType.BotDM,
        InteractionContextType.PrivateChannel
    ];
    /**
     * The components that the command is able to use.
     * You pass these here so the handler can listen for them..
     */
    components = [];
    /**
     * All the modals that the command is able to use.
     * You pass these here so the handler can listen for them.
     */
    modals = [];
    /**
     * Serializes the command into a JSON object that can be sent to Discord
     * @internal
     */
    serialize() {
        if (this.type === ApplicationCommandType.PrimaryEntryPoint) {
            throw new Error("Primary Entry Point commands cannot be serialized");
        }
        // Only chat input commands can have descriptions
        if (this.type === ApplicationCommandType.ChatInput) {
            const data = {
                name: this.name,
                type: this.type,
                description: this.description,
                options: this.serializeOptions(),
                integration_types: this.integrationTypes,
                contexts: this.contexts
            };
            return data;
        }
        const data = {
            name: this.name,
            type: this.type,
            options: this.serializeOptions(),
            integration_types: this.integrationTypes,
            contexts: this.contexts
        };
        return data;
    }
}
