// @buape/carbon-request/dist/src/RequestClient.js
import { DiscordError } from "./errors/DiscordError.js";
import { RateLimitError } from "./errors/RatelimitError.js";
const defaultOptions = {
    tokenHeader: "Bot",
    baseUrl: "https://discord.com/api",
    apiVersion: 10,
    userAgent: "DiscordBot (https://github.com/buape/carbon, v0.0.0)",
    timeout: 15000,
    queueRequests: true
};
/**
 * This is the main class that handles making requests to the Discord API.
 * It is used internally by Carbon, and you should not need to use it directly, but feel free to if you feel like living dangerously.
 */
export class RequestClient {
    /**
     * The options used to initialize the client
     */
    options;
    queue = [];
    token;
    rateLimitResetTime;
    abortController = null;
    constructor(token, options) {
        this.token = token;
        this.options = {
            ...defaultOptions,
            ...options
        };
        this.rateLimitResetTime = 0;
    }
    async waitForRateLimit() {
        const delay = this.rateLimitResetTime - Date.now();
        if (delay > 0) {
            await new Promise((resolve) => setTimeout(resolve, delay));
        }
    }
    async get(path) {
        return await this.request("GET", path);
    }
    async post(path, data) {
        return await this.request("POST", path, data);
    }
    async patch(path, data) {
        return await this.request("PATCH", path, data);
    }
    async put(path, data) {
        return await this.request("PUT", path, data);
    }
    async delete(path, data) {
        return await this.request("DELETE", path, data);
    }
    async request(method, path, data) {
        if (this.options.queueRequests) {
            return new Promise((resolve, reject) => {
                this.queue.push({ method, path, data, resolve, reject });
                this.processQueue();
            });
        }
        return new Promise((resolve, reject) => {
            this.executeRequest({ method, path, data, resolve, reject })
                .then(resolve)
                .catch((err) => {
                reject(err);
            });
        });
    }
    async executeRequest(request) {
        await this.waitForRateLimit();
        const { method, path, data } = request;
        const url = `${this.options.baseUrl}${path}`;
        const headers = new Headers({
            Authorization: `${this.options.tokenHeader} ${this.token}`,
            ...(data?.files ? {} : { "Content-Type": "application/json" })
        });
        this.abortController = new AbortController();
        let body;
        if (data?.files != null) {
            const formData = new FormData();
            for (const [index, file] of data.files.entries()) {
                let { data: fileData } = file;
                if (!(fileData instanceof Blob)) {
                    fileData = new Blob([fileData]);
                }
                formData.append(`files[${file.id ?? index}]`, fileData, file.name);
            }
            if (data.body != null) {
                formData.append("payload_json", JSON.stringify(data.body));
            }
            body = formData;
        }
        else if (data?.body != null) {
            if (data.rawBody) {
                body = data.body;
            }
            else {
                body = JSON.stringify(data.body);
            }
        }
        const response = await fetch(url, {
            method,
            headers,
            body,
            signal: this.abortController.signal
        });
        if (response.status === 429) {
            const responseBody = await response.json();
            const rateLimitError = new RateLimitError(response, responseBody);
            if (this.options.queueRequests) {
                const rateLimitResetTime = Number(response.headers.get("Retry-After")) * 1000;
                this.rateLimitResetTime = Date.now() + rateLimitResetTime;
                if (rateLimitError.scope === "global") {
                    await new Promise((res) => setTimeout(res, rateLimitResetTime));
                    this.queue.unshift(request);
                }
            }
            throw rateLimitError;
        }
        if (response.status >= 400 && response.status < 600) {
            throw new DiscordError(response, await response.json());
        }
        try {
            return await response.json();
        }
        catch (err) {
            if (err instanceof SyntaxError) {
                // If there is no JSON
                return await response.text();
            }
            throw err;
        }
    }
    async processQueue() {
        if (this.queue.length === 0)
            return;
        const queueItem = this.queue.shift();
        if (!queueItem)
            return;
        const { method, path, data, resolve, reject } = queueItem;
        try {
            const result = await this.executeRequest({
                method,
                path,
                data,
                resolve,
                reject
            });
            resolve(result);
        }
        catch (error) {
            if (error instanceof RateLimitError &&
                this.options.queueRequests &&
                error.scope === "global") {
                this.queue.unshift(queueItem);
            }
            else {
                if (error instanceof Error) {
                    reject(error);
                }
                else {
                    reject(new Error("Unknown error during request", { cause: error }));
                }
            }
        }
        finally {
            this.abortController = null;
            this.processQueue();
        }
    }
    abortAllRequests() {
        if (this.abortController) {
            this.abortController.abort();
        }
        this.queue = [];
    }
}
