// @buape/carbon/dist/src/internals/OptionsHandler.js
import { ApplicationCommandOptionType, Routes } from "discord-api-types/v10";
import { Base } from "../abstracts/Base.js";
import { Role, User, channelFactory } from "../index.js";
/**
 * This class is used to parse the options of a command, and provide errors for any missing or invalid options.
 * It is used internally by the Command class.
 */
export class OptionsHandler extends Base {
    /**
     * The raw options that were in the interaction data, before they were parsed.
     */
    raw;
    constructor(client, options) {
        super(client);
        this.raw = [];
        for (const option of options) {
            if (option.type === ApplicationCommandOptionType.Subcommand) {
                for (const subOption of option.options ?? []) {
                    this.raw.push(subOption);
                }
            }
            else if (option.type === ApplicationCommandOptionType.SubcommandGroup) {
                for (const subOption of option.options ?? []) {
                    if (subOption.options) {
                        for (const subSubOption of subOption.options ?? []) {
                            this.raw.push(subSubOption);
                        }
                    }
                }
            }
            else {
                this.raw.push(option);
            }
        }
    }
    getString(key, required = false) {
        const value = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.String)?.value;
        if (required) {
            if (!value || typeof value !== "string")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!value || typeof value !== "string")
            return undefined;
        return value;
    }
    getInteger(key, required = false) {
        const value = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.Integer)?.value;
        if (required) {
            if (!value || typeof value !== "number" || !Number.isSafeInteger(value))
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!value ||
            typeof value !== "number" ||
            !Number.isSafeInteger(value))
            return undefined;
        return value;
    }
    getNumber(key, required = false) {
        const value = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.Number)?.value;
        if (required) {
            if (!value || typeof value !== "number")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!value || typeof value !== "number")
            return undefined;
        return value;
    }
    getBoolean(key, required = false) {
        const value = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.Boolean)?.value;
        if (required) {
            if (!value || typeof value !== "boolean")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!value || typeof value !== "boolean")
            return undefined;
        return value;
    }
    getUser(key, required = false) {
        const id = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.User)?.value;
        if (required) {
            if (!id || typeof id !== "string")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!id || typeof id !== "string")
            return undefined;
        return new User(this.client, id);
    }
    async getChannel(key, required = false) {
        const id = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.Channel)?.value;
        if (required) {
            if (!id || typeof id !== "string")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!id || typeof id !== "string")
            return undefined;
        const data = (await this.client.rest.get(Routes.channel(id)));
        return channelFactory(this.client, data);
    }
    getRole(key, required = false) {
        const id = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.Role)?.value;
        if (required) {
            if (!id || typeof id !== "string")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!id || typeof id !== "string")
            return undefined;
        return new Role(this.client, id);
    }
    async getMentionable(key, required = false) {
        const id = this.raw.find((x) => x.name === key && x.type === ApplicationCommandOptionType.Mentionable)?.value;
        if (required) {
            if (!id || typeof id !== "string")
                throw new Error(`Missing required option: ${key}`);
        }
        else if (!id || typeof id !== "string")
            return undefined;
        try {
            const user = new User(this.client, id);
            await user.fetch();
            return user;
        }
        catch {
            return new Role(this.client, id);
        }
    }
}
