// discord-verify/dist/shared/discord-verify.f36cec27.mjs
const encoder = new TextEncoder();
const KEYS = {
  ZERO: 48,
  A: 65,
  a: 97
};
function hexCharToBinary(char) {
  const code = char.charCodeAt(0);
  if (code >= KEYS.a) {
    return code - KEYS.a + 10;
  }
  if (code >= KEYS.A) {
    return code - KEYS.A + 10;
  }
  return code - KEYS.ZERO;
}
function hexStringToBinary(key) {
  if (key == null || key.length % 2 !== 0) {
    return new Uint8Array(0).buffer;
  }
  const view = new Uint8Array(key.length / 2);
  for (let i = 0, o = 0; i < key.length; i += 2, ++o) {
    view[o] = hexCharToBinary(key[i]) << 4 | hexCharToBinary(key[i + 1]);
  }
  return view.buffer;
}
async function getCryptoKey(publicKey, subtleCrypto, algorithm) {
  const key = await subtleCrypto.importKey(
    "raw",
    hexStringToBinary(publicKey),
    algorithm,
    true,
    ["verify"]
  );
  return key;
}
const PlatformAlgorithm = {
  Web: "Ed25519",
  /**
   * Node v18.4.0+
   * Node v16.17.0+
   * For Node v17, use OldNode
   */
  NewNode: "Ed25519",
  /**
   * Node v18.3.0 and below
   * Node v17.0.0+
   * Node v16.16.0 and below
   */
  OldNode: {
    name: "NODE-ED25519",
    namedCurve: "NODE-ED25519",
    public: true
  },
  Cloudflare: {
    name: "NODE-ED25519",
    namedCurve: "NODE-ED25519",
    public: true
  },
  /**
   * Despite being documented as 	`{ name: "eddsa", namedCurve: "ed25519"}` or
   * `{ name: "ecdsa", namedCurve: "ed25519" }`, Vercel uses the same format as
   * Cloudflare in Production (despite Dev using documented formats)
   */
  VercelProd: {
    name: "NODE-ED25519",
    namedCurve: "NODE-ED25519",
    public: true
  },
  /**
   * Despite being documented as using this format, Vercel uses the same format
   * as Cloudflare in Production and only uses this format in Development.
   */
  VercelDev: { name: "eddsa", namedCurve: "ed25519" }
};
async function isValidRequest(request, publicKey, subtleCrypto, algorithm = PlatformAlgorithm.NewNode, expirationTime = 15 * 6e4) {
  const clone = request.clone();
  const timestamp = clone.headers.get("X-Signature-Timestamp");
  const signature = clone.headers.get("X-Signature-Ed25519");
  const body = await clone.text();
  return verify(
    body,
    signature,
    timestamp,
    publicKey,
    subtleCrypto,
    algorithm,
    expirationTime
  );
}
async function verify(rawBody, signature, timestamp, publicKey, subtleCrypto, algorithm = PlatformAlgorithm.NewNode, expirationTime = 15 * 6e4) {
  if (timestamp == null || signature == null || rawBody == null) {
    return false;
  }
  if (Math.abs(Date.now() / 1e3 - parseInt(timestamp, 10)) > expirationTime) {
    return false;
  }
  const key = await getCryptoKey(publicKey, subtleCrypto, algorithm);
  const name = typeof algorithm === "string" ? algorithm : algorithm.name;
  const isVerified = await subtleCrypto.verify(
    name,
    key,
    hexStringToBinary(signature),
    encoder.encode(`${timestamp ?? ""}${rawBody}`)
  );
  return isVerified;
}

export { PlatformAlgorithm as P, hexStringToBinary as h, isValidRequest as i, verify as v };
